### ISSUE TITLE
Newruntime: incomplete treatment of `owned` non-`ref`'s, including documentation...

### ISSUE THREAD

**Issue by GordonBGood** [nim-lang/Nim#12021]
Although recently updated and improved, [the documentation for the newruntime destructors](https://github.com/nim-lang/Nim/blob/devel/doc/destructors.rst) still doesn't mention that the `owned` modifier can also be applied to any type and not just `ref`'s and what that implies.

It is not documented that currently, non-global `proc`'s (which may be closures capturing external bindings with a heap allocated environment reference) are automatically elevated to `owned` status (unless one takes special measures to apply the `unown` template when they are created); in fact, this elevation of `proc`'s (including closures) to `owned` status is not consistently internally used as per issue #12008 

The documentation should be updated to reflect the implementation, with exact specification that such elevation of status to an `owned` type does not implicitly ref count (as these types are not allocated to the heap or are primitive pointers that are outside the automatic memory management).

### Example
```nim
# test newruntime `owned` anything...

# compile with "--newruntime"

template own[T](x: T): owned T = # is the opposite of `unown` template!
  (proc (): owned T = cast[ptr T](result.addr)[] = x)()

proc main() =
  let x = own 42; x.typeof.echo # showing we have a `owned int`!= 
  let y = unown x; y.typeof.echo # must `unown` as not last use of `x`;
  (unown x).echo # last use of `x`, must be unowned here in order to print
  # there is no `$` for `owned` types to automatically print without `unown`'ing them!

main()
```

### Current Output
```
owned int
int
42
0
```
The above behaviour is correct, just not documented fully, as without the use of `unown` on the assignment of the `owned int` (`x`), the following error is obtained:
```
Error: '=' is not available for type <owned int>; requires a copy because it's not the last read of 'x'; another read is done here:
```
This shows that `owned` of other types follows the rule of "no copying of `owned` types, only moving is possible" correctly.

### Expected Output
This is as expected, but undocumented that it is possible and that the `owned` rules can  be applied to any type if they are forced to be `owned`.

### Additional Information
```
$ nim -v
Nim Compiler Version 0.20.99 DEVEL nightly build as of 02-August-2019 on Linux and Windows 64-bit
```

**Comment by Araq** [comment]
Actually ``owned int`` is not a valid type and the compiler should prevent it. ``owned`` is only applicable to ``proc``, ``ref`` and ``ptr``.

**Comment by GordonBGood** [comment]
@Araq, the issue still stands though, there doesn't seem to be any documentation on what can be `owned`, with considerations as follows:
1. Obviously we have `owned ref` as that was the whole start of our use of the idea from Bacon/Dingle (B/D).
2. Since the semantics between `ref` and `ptr` are blurred by the `newruntime` in that they would seem to be the same thing, I kind of see why you would want `ptr` able to be `owned`, too, but there isn't a lot of point as (currently) `owned ptr`'s don't have the automatic destruction semantics applied, so being owned does nothing more than for an `owned int`: it prevents copying unless it is "last read of" in which case it is a "move".
3. I see why you've made closures to be `owned` in that they contain an `owned` `ptr` or `ref` to the captured environment, although that is causing some problems currently as to nested closures as in my issue #12008 and makes it more complex to obtain another owned closure with the same environment (seemly must use a closure factory as used in the sample code in that issue).
4. Given that closures can be owned as above, and given than "{.nimcall.}"`proc`'s can be assigned to closures, I kind of see why you've made these `proc`'s `owned` as well, although being `owned` does nothing more for ordinary `proc`'s than it does for `owned int`.  
5. One could make a case for having `owned seq`'s and `owned string`'s (or some new version as to actual structure thereof) as they contain heap `ref`'s or `ptr`'s that must be controlled.  Given the number of current issues on `newruntime` `seq`'s in dealing with "special cases" as in `shallow`/`shallowCopy`, etc., this might be a cleaner way to handle it?  If this is possible, it makes an new type (say `table`) that contains references to heap data and implemented in the same `owned` way more consistent as to avoiding "special cases" to make them seem part of the same system - the system is already there and it is `owned`.
6. So, why not allow anything to be `owned`, since the compiler already handles this and there would likely have to be "special case" code in the compiler to prevent it?  It would open to door to an eventual "affine" type safety system as follows: if everything were first created `owned`, then it would be reasonably easy to add a "singly mutably referenced" check to all `owned` in a similar way that reference counting is added to `owned ref` (and eventually `owned ptr`?) so that an `owned` that is mutable (`var`) can have at most one lent out mutable reference at a time,, and if loaned out cannot be mutated through the `owned` binding but only through the singly loaned mutable binding until the loan ends.  I am not suggesting that Nim version 2 would have this, but it would open the door to it for Nim version 3.  If this were adopted, Nim would be moving toward a "type safe" system something like Rust's but much simpler, although only considering "static" lifetimes and not the multiple levels of lifetimes as does Rust.

In conclusion, given that it seems we already have `owned anything`, why work at preventing it when it doesn't seem to do any harm and there may be (future and/or present) advantages?

**Comment by Araq** [comment]
> In conclusion, given that it seems we already have owned anything, why work at preventing it when it doesn't seem to do any harm and there may be (future and/or present) advantages?

Good points, in addtion to that ``owned`` can creep into generic code:

```nim

proc create[T](): owned T

create[int]()
```

However, the compiler isn't ready for dealing with ``owned`` everywhere.

**Comment by ringabout** [comment]
newruntime is deprecated and taken over by ARC/ORC.
